src/astar.rs

```diff
@@ -0,0 +1,168 @@
+use std::collections::{
+    HashMap,
+    BinaryHeap,
+};
+use std::collections::hash_map::Entry::{
+    Occupied,
+    Vacant,
+};
+
+use std::hash::Hash;
+
+use scored::MinScored;
+use super::visit::{
+    EdgeRef,
+    GraphBase,
+    IntoEdges,
+    VisitMap,
+    Visitable,
+};
+
+use algo::Measure;
+
+/// [Generic] A* shortest path algorithm.
+///
+/// Compute the length of the shortest path from `start` to `finish`.
+///
+/// `finish` is implicitly given via the `is_goal` callback, which should return `true` if the
+/// given node is the finish node.
+///
+/// The function `edge_cost` should return the cost for a particular edge, which is used to compute
+/// path costs.
+///
+/// The function `estimate_cost` should return the estimated cost to the finish for a particular
+/// node, also used to compute path costs.
+///
+/// ```
+/// use petgraph::Graph;
+/// use petgraph::algo::astar;
+///
+/// let mut g = Graph::new();
+/// let a = g.add_node((0., 0.));
+/// let b = g.add_node((2., 0.));
+/// let c = g.add_node((1., 1.));
+/// let d = g.add_node((0., 2.));
+/// let e = g.add_node((3., 3.));
+/// let f = g.add_node((4., 2.));
+/// let ab = g.add_edge(a, b, 2);
+/// let ad = g.add_edge(a, d, 4);
+/// let bc = g.add_edge(b, c, 1);
+/// let bf = g.add_edge(b, f, 7);
+/// let ce = g.add_edge(c, e, 5);
+/// let ef = g.add_edge(e, f, 1);
+/// let de = g.add_edge(d, e, 1);
+///
+/// let path = astar(&g, a, |finish| finish == f, |e| *e.weight(), |_| 0);
+/// assert_eq!(path, Some(vec![a, d, e, f]));
+/// ```
+///
+/// The graph should be `Visitable` and implement `IntoEdges`. Edge costs must be non-negative.
+///
+/// Returns a Path of subsequent `NodeId` from start to finish, if one was found.
+pub fn astar<G, F, H, K, IsGoal>(graph: G, start: G::NodeId, is_goal: IsGoal,
+                                     mut edge_cost: F, mut estimate_cost: H)
+    -> Option<Path<G>>
+    where G: IntoEdges + Visitable,
+          IsGoal: Fn(G::NodeId) -> bool,
```

Comment:
any reason is goal is Fn instead of FnMut?

Reply:
Nice catch, I'll switch that.